### PR TITLE
feat(Rewards): add mintBatchReward

### PR DESCRIPTION
**Pull request by aliXsed**
In this implementation I have defined the struct BatchReward as:
```
    struct BatchReward {
        address[] recipients;
        uint256[] amounts;
        uint256 sequence;
    }
```
assuming `recipients[i]` corresponds to `amounts[i]`. This means we expect `recipients.length == amounts.length`. That's why `mintBatchReward` needed to check this assumption. The reason I have gone for this design was to avoid using **nested types** which could make `digestReward` more complicated and costly. `digestReward` is needed for checking the signature.

### CODE REVIEW

File: src/Rewards.sol [hunk 1]
```diff
@@ -159,30 +184,78 @@ contract Rewards is AccessControl, EIP712 {
...
+        uint256 batchSum = _batchSum(batch);
+
+        _checkedUpdateClaimed(batchSum);
+
+        // Safe to increment the sequence after checking this is the expected number (no overflow for the age of universe even with 1000 reward claims per second)
+        batchSequence = batch.sequence + 1;
+
+        for (uint256 i = 0; i < batch.recipients.length; i++) {
+            nodl.mint(batch.recipients[i], batch.amounts[i]);
+        }
```
Comment (ETeissonniere) [inline src/Rewards.sol:226]: We can avoid the double for loop here (other loop in `_batchSum`) by rewriting the for iteration and add one last revert statement afterwards

**Review by ETeissonniere — COMMENTED** [review]
No real comments, just whitespacing (we don't need one space between every line :D) and one open question on a double for loop which might be optimizable

File: src/Rewards.sol (hunk 1 above)
Comment (aliXsed) [reply src/Rewards.sol]: I tested the potential gas saving in your suggestion and found it either absolute zero or very negligible (less than 0.1%). While checking first and minting later has the benefit of early failure (gas saving there) and follows the best practice of updating state first before calling into any other contract.

Comment (ETeissonniere) [reply src/Rewards.sol]: Makes sense, let's keep it as is

**Review by ETeissonniere — APPROVED** [review]

**Pull request merged**